core/src/main/java/org/apache/iceberg/ManifestGroup.java

```diff
@@ -234,7 +255,13 @@ private <T> Iterable<CloseableIterable<T>> entries(
           manifest -> manifest.hasAddedFiles() || manifest.hasDeletedFiles());
     }
 
-    matchingManifests = Iterables.filter(matchingManifests, manifestPredicate::test);
+    matchingManifests = Iterables.filter(matchingManifests, manifest -> {
```

Comment:
Rather than changing the filtering, why not make a new Iterable that updates a counter? Then you'd have less custom code:

```java
  class <T> Iterable<T> count(Counter counter, Iterable<T> iterable) {
    return () -> new Iterator<T> {
      Iterator<T> iter = iterable.iterator();

      public boolean hasNext() {
        return iter.hasNext();
      }

      public T next() {
        T next = iter.next();
        counter.increment();
        return next;
      }
    }
  }

  ...

  matchingManifests = count(scanMetrics.matchingDataManifests(), Iterables.filter(...));
```

Reply:
added that functionality